### PR TITLE
interfaces_assign.php UI speedup as requested on forum bounty board

### PR DESCRIPTION
**Pull request by loonylion**
Improves significantly and hopefully fixes bug#6400

Refactored interfaces_assign.php to to benefit people with large numbers of VLANs, as requested on the forum at https://forum.pfsense.org/index.php?topic=137391.0. Also contains a minor speedup for interfaces_vlan.php. Modified functions are contained in interfaces_fast.inc. Profiling code is still present but commented out, as is replaced code.

### CODE REVIEW

**Comment by loonylion** [comment]
New commit also removes some commented code, some of which appeared to be slowing down page load significantly. Page load times are now much closer to generation times than previously.

**Comment by sbeaver-netgate** [comment]
Incorrect license text in new file

**Comment by loonylion** [comment]
licence text changed as requested

File: src/etc/inc/interfaces_fast.inc [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+ *
+ * Functions in this file may not require/use parameters when called, however 
+ * in most cases they will accept parameters as per their original forms, for 
+ * consistency.
+ 
+ */
+require_once('interfaces.inc');
+/*
+* does_interface_exist_fast
+Returns an array of interrfaces which exist on the system.
```
Comment (sbeaver-netgate) [inline src/etc/inc/interfaces_fast.inc]: Spelling: interrfaces

File: src/etc/inc/interfaces_fast.inc [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+	$ints = get_interface_arr($flush);
+	return array_flip($ints);
+}
+
+/*
+ * convert_real_interface_to_friendly_interface_name_fast($interface): convert fxp0 -> wan, etc.
+ Returns an array of interfaces and friendly names.
+ Doesn't actually use any parameters, they're just accepted for consistency with the function it replaces
+ */
```
Comment (sbeaver-netgate) [inline src/etc/inc/interfaces_fast.inc]: I don't think this consistency is required, and it rather obfuscates the code. Could these arguments be removed wherever used?

File: src/usr/local/www/interfaces_assign.php [hunk 3]
```diff
@@ -365,8 +382,13 @@ function interface_assign_description($portinfo, $portname) {
 				}
 			}
 		}
+		/*$timeb = microtime(true);
+		$profile['post_build_config'] = $timeb-$timea;
 
+		$timea = microtime(true);*/
 		write_config();
+		/*$timeb = microtime(true);
+		$profile['post_write_config']=$timeb-$timea;*/
```
Comment (sbeaver-netgate) [inline src/usr/local/www/interfaces_assign.php]: Let's just remove stuff that is no longer required.

File: src/usr/local/www/interfaces_assign.php [hunk 4]
```diff
@@ -441,18 +472,22 @@ function interface_assign_description($portinfo, $portname) {
...
-		$unused_portlist[$portname] = $portinfo;
-	}
-}
+//$timea = microtime(true);
+$portArray = array_keys($portlist);
+
+/*  this code scales much much better 
+0.0065770149230957 seconds
+vs
+0.49271988868713 seconds with 400 vlans*/
```
Comment (sbeaver-netgate) [inline src/usr/local/www/interfaces_assign.php]: Comment not useful here

File: src/usr/local/www/interfaces_assign.php [hunk 5]
```diff
@@ -501,6 +536,32 @@ function interface_assign_description($portinfo, $portname) {
...
+/*generate the port select box only once. 
+Not indenting the HTML and keeping each option to one line in
+this function results in HTML code that is 8.9KB smaller with 500
+VLANS than the original code structure. Multiplied by 500 VLANs 
+this means the page is ~4.5MB smaller and takes over 1s less to 
+transmit and also renders significantly faster in the browser.
+This is too much of an improvement to ignore. Tested with 500
+VLANS, total page output is 17,197.33KB with 8456ms spent 
+waiting for page generation and transmission, compared to 
+21,697.89KB with a wait of 9734ms*/
```
Comment (sbeaver-netgate) [inline src/usr/local/www/interfaces_assign.php]: This is a self-congratulatory comment.  "Keep HTML options on one line to minimize size/load-time" would suffice.

File: src/usr/local/www/interfaces_assign.php [hunk 6]
```diff
@@ -525,23 +588,31 @@ function interface_assign_description($portinfo, $portname) {
...
-<?php foreach ($portlist as $portname => $portinfo):?>
-					<option value="<?=$portname?>" <?=($portname == $iface['if']) ? ' selected': ''?>>
-						<?=interface_assign_description($portinfo, $portname)?>
-					</option>
-<?php endforeach;?>
+<?php //$timea3 = microtime(true);
+/* replacing the port select menu generation loop that has count(interfaces) iterations 
+and is run count(interfaces) times with a pre-prepared select menu generated outside of
+this loop has produced a significant improvement in page generation and load time */
+//foreach ($portlist as $portname => $portinfo):
```
Comment (sbeaver-netgate) [inline src/usr/local/www/interfaces_assign.php]: A more concise comment would be nice.

File: src/usr/local/www/interfaces_assign.php [hunk 7]
```diff
@@ -550,17 +621,23 @@ function interface_assign_description($portinfo, $portname) {
...
-<?php foreach ($unused_portlist as $portname => $portinfo):?>
-					<option value="<?=$portname?>" <?=($portname == $iface['if']) ? ' selected': ''?>>
-						<?=interface_assign_description($portinfo, $portname)?>
-					</option>
-<?php endforeach;?>
+<?php //$timea2 = microtime(true);
+/* As with the gettext() calls, I've removed the interface_assign_description() calls and
+replaced them with my own interface_assign_description_fast() function that's called once
+outside of the loop. Also like the gettext() edits, this change keeps paying for itself.
+ Also, removing the indents and newlines saves potentially ~9KB of HTML with 500 unassigned VLANs */
```
Comment (sbeaver-netgate) [inline src/usr/local/www/interfaces_assign.php]: Comment not helpful here. "HTML indenting not used to save tmie/space" would suffice.

**Comment by loonylion** [comment]
changes made, the long comments were intended to explain my reasoning behind some of the more significant changes or changes that appeared to possibly conflict with the coding style guide (in the case of the indent removals)

**Comment by sbeaver-netgate** [comment]
Thanks. - Merged.
Those comments are better here than in the source I think :)

**Comment by sbeaver-netgate** [comment]
Appears to be causing nnumerous issues: https://redmine.pfsense.org/issues/8223

**Comment by sbeaver-netgate** [comment]
PR was reverted due to errors introduced.

**Comment by loonylion** [comment]
Both bugs should be fixed now. I reverted back to stock pfSense code for naming OPT interfaces since the improvement wasn't huge anyway, which should take care of bug #8222 and I corrected the broken reference causing bug #8223

**Pull request merged**